File: LV 2/큰 수 만들기.py
from collections import deque

def solution(number, k):
    stack = deque()
    delcnt = 0
    for i, n in enumerate(number):
        while stack and delcnt != k:
            if stack[-1] < n:
                delcnt += 1
                stack.pop()
            else:
                break
        stack.append(n)
    
    answer = "".join(stack)
    if delcnt < k: # 충분히 지우지 못한 경우 뒤에서부터 지워줌.
        answer = answer[:-(k-delcnt)]
        
    return answer

File: LV 2/큰 수 만들기.java
import java.util.Stack;

class Solution {
    int changeCount;

    public int[] toIntArray(String s) {
        return s.chars()
                .map(ch -> ch - '0')
                .toArray();
    }

    public void push(int number, int limit, Stack<Integer> stack) {
        while (!stack.isEmpty() && stack.peek() < number && changeCount < limit) {
            stack.pop();
            changeCount++;
        }
        stack.push(number);
    }

    public String makeNumber(Stack<Integer> stack, int limit) {
        int totalCount = stack.size() - (limit - changeCount);
        String s = "";
        for (int i = 0; i < totalCount; i++) {
            s += stack.get(i);
        }
        return s;
    }

    public String solution(String number, int k) {
        String answer = "";
        Stack<Integer> stack = new Stack<Integer>();
        int[] numberArray = toIntArray(number);
        changeCount = 0;
        for (int eachNumber : numberArray) {
            push(eachNumber, k, stack);
        }
        return makeNumber(stack, k);
    }
}
